### ISSUE TITLE
SIGSEV in Bindings/SetValueImpl

### ISSUE THREAD

**Issue by Noires** [MvvmCross/MvvmCross#1379]
In my project i stumbled on some errors with the bindings (http://stackoverflow.com/questions/37585394/mvxnotifypropertychanged-application-crash-on-setproperty). The error occured for me in _MvxPropertyInfoTargetBinding_ and _MvxTextViewTextTargetBinding_, but might be able to occur in other binding classes aswell. 

Could only imagine it sometimes happens that the reference gets cleared while the value gets set? Dunno.

### Configuration

**Version:** 4.2.0

**Platform:** Android

```
backtrace:
    #00 pc 00000000001b1524  /system/lib64/libskia.so (SkPaint::operator=(SkPaint const&)+84)
    #01 pc 000000000010ebb8  /system/lib64/libandroid_runtime.so (android::Paint::operator=(android::Paint const&)+24)
    #02 pc 00000000000f1225  /data/dalvik-cache/x86_64/system@framework@boot.oat

 Stacktrace:
   at <unknown> <0xffffffff>
   at (wrapper managed-to-native) Java.Interop.NativeMethods.java_interop_jnienv_call_nonvirtual_void_method_a (intptr,intptr&,intptr,intptr,intptr,Java.Interop.JniArgumentValue*) <IL 0x00010, 0x000bf>
   at Java.Interop.JniEnvironment/InstanceMethods.CallNonvirtualVoidMethod (Java.Interop.JniObjectReference,Java.Interop.JniObjectReference,Java.Interop.JniMethodInfo,Java.Interop.JniArgumentValue*) [0x0008f] in /Users/builder/data/lanes/3340/4e275588/source/Java.Interop/src/Java.Interop/Java.Interop/JniEnvironment.g.cs:12079
   at Java.Interop.JniPeerMembers/JniInstanceMethods.InvokeNonvirtualVoidMethod (string,Java.Interop.IJavaPeerable,Java.Interop.JniArgumentValue*) [0x00022] in /Users/builder/data/lanes/3340/4e275588/source/Java.Interop/src/Java.Interop/Java.Interop/JniPeerMembers.JniInstanceMethods_Invoke.cs:41
   at Android.Widget.TextView.set_TextFormatted (Java.Lang.ICharSequence) [0x0001e] in /Users/builder/data/lanes/3340/4e275588/source/monodroid/src/Mono.Android/platforms/android-23/src/generated/Android.Widget.TextView.cs:2787
   at Android.Widget.TextView.set_Text (string) [0x00013] in /Users/builder/data/lanes/3340/4e275588/source/monodroid/src/Mono.Android/platforms/android-23/src/generated/Android.Widget.TextView.cs:2798
   at MvvmCross.Binding.Droid.Target.MvxTextViewTextTargetBinding.SetValueImpl (object,object) <IL 0x00040, 0x00313>
   at MvvmCross.Binding.Bindings.Target.MvxConvertingTargetBinding.SetValue (object) <IL 0x0008b, 0x00586>
   at MvvmCross.Binding.Bindings.MvxFullBinding/<>c__DisplayClass23_0.<UpdateTargetFromSource>b__0 () <IL 0x00020, 0x00158>
   at MvvmCross.Droid.Views.MvxAndroidMainThreadDispatcher.RequestMainThreadAction (System.Action) <IL 0x0001f, 0x001a3>
   at MvvmCross.Binding.Bindings.MvxFullBinding.UpdateTargetFromSource (object,System.Threading.CancellationToken) <IL 0x00061, 0x0043e>
   at MvvmCross.Binding.Bindings.MvxFullBinding.<CreateSourceBinding>b__19_0 (object,System.EventArgs) <IL 0x0001b, 0x00183>
   at MvvmCross.Binding.Bindings.SourceSteps.MvxSourceStep.SendSourcePropertyChanged () <IL 0x00011, 0x000ce>
   at MvvmCross.Binding.Bindings.SourceSteps.MvxCombinerSourceStep.SubStepOnChanged (object,System.EventArgs) <IL 0x00001, 0x0007f>
   at MvvmCross.Binding.Bindings.SourceSteps.MvxSourceStep.SendSourcePropertyChanged () <IL 0x00011, 0x000ce>
   at MvvmCross.Binding.Bindings.SourceSteps.MvxCombinerSourceStep.SubStepOnChanged (object,System.EventArgs) <IL 0x00001, 0x0007f>
   at MvvmCross.Binding.Bindings.SourceSteps.MvxSourceStep.SendSourcePropertyChanged () <IL 0x00011, 0x000ce>
   at MvvmCross.Binding.Bindings.SourceSteps.MvxCombinerSourceStep.SubStepOnChanged (object,System.EventArgs) <IL 0x00001, 0x0007f>
   at MvvmCross.Binding.Bindings.SourceSteps.MvxSourceStep.SendSourcePropertyChanged () <IL 0x00011, 0x000ce>
   at MvvmCross.Binding.Bindings.SourceSteps.MvxPathSourceStep.SourceBindingOnChanged (object,System.EventArgs) <IL 0x00001, 0x0006f>
   at MvvmCross.Binding.Bindings.Source.MvxSourceBinding.FireChanged () <IL 0x00011, 0x000ce>
   at MvvmCross.Binding.Bindings.Source.Leaf.MvxLeafPropertyInfoSourceBinding.OnBoundPropertyChanged () <IL 0x00001, 0x00067>
   at MvvmCross.Binding.Bindings.Source.MvxPropertyInfoSourceBinding.SourcePropertyChanged (object,System.ComponentModel.PropertyChangedEventArgs) <IL 0x00021, 0x001a1>
   at (wrapper runtime-invoke) <Module>.runtime_invoke_void__this___object_object (object,intptr,intptr,intptr) <IL 0x0005d, 0x00146>
   at <unknown> <0xffffffff>
   at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&) <IL 0x00016, 0x000b4>
   at System.Reflection.MonoMethod.Invoke (object,System.Reflection.BindingFlags,System.Reflection.Binder,object[],System.Globalization.CultureInfo) [0x00038] in /Users/builder/data/lanes/2923/52635947/source/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:295
   at System.Reflection.MethodBase.Invoke (object,object[]) [0x00000] in /Users/builder/data/lanes/2923/52635947/source/mono/external/referencesource/mscorlib/system/reflection/methodbase.cs:229
   at MvvmCross.Platform.WeakSubscription.MvxWeakEventSubscription`2<TSource_REF, TEventArgs_REF>.OnSourceEvent (object,TEventArgs_REF) [0x0000f] in C:\vcs\git\MvvmCross\MvvmCross\Platform\Platform\WeakSubscription\MvxWeakEventSubscription.cs:75
   at (wrapper delegate-invoke) <Module>.invoke_void_object_PropertyChangedEventArgs (object,System.ComponentModel.PropertyChangedEventArgs) <IL 0x00073, 0x00151>
   at MvvmCross.Core.ViewModels.MvxNotifyPropertyChanged/<>c__DisplayClass10_0.<RaisePropertyChanged>b__0 () <IL 0x0001c, 0x000eb>
   at MvvmCross.Droid.Views.MvxAndroidMainThreadDispatcher.RequestMainThreadAction (System.Action) <IL 0x0001f, 0x001a3>
   at MvvmCross.Platform.Core.MvxMainThreadDispatchingObject.InvokeOnMainThread (System.Action) [0x00000] in C:\vcs\git\MvvmCross\MvvmCross\Platform\Platform\Core\MvxMainThreadDispatchingObject.cs:18
   at MvvmCross.Core.ViewModels.MvxNotifyPropertyChanged.RaisePropertyChanged (System.ComponentModel.PropertyChangedEventArgs) <IL 0x00044, 0x00347>
   at MvvmCross.Core.ViewModels.MvxNotifyPropertyChanged.RaisePropertyChanged (string) <IL 0x00009, 0x000e4>
   at MvvmCross.Core.ViewModels.MvxNotifyPropertyChanged.RaisePropertyChanged<int> (System.Linq.Expressions.Expression`1<System.Func`1<int>>) <IL 0x0000a, 0x000bb>
   at XCPCore.Models.Base.BaseCategory.set_Count (int) [0x00008] in W:\DevWA\MobileXCP~BARONDST\MobileXCP\MobileXcp\MobileXcp.Core\Models\Base\BaseCategory.cs:59
   at XCPCore.ViewModels.Cal.CalViewModel.UpdateChannels () [0x0007c] in W:\DevWA\MobileXCP~BARONDST\MobileXCP\MobileXcp\MobileXcp.Core\ViewModels\Cal\CalViewModel.cs:177
   at XCPCore.ViewModels.Cal.BaseCalViewModel.Refresh () [0x00001] in W:\DevWA\MobileXCP~BARONDST\MobileXCP\MobileXcp\MobileXcp.Core\ViewModels\Cal\BaseCalViewModel.cs:91
   at MobileXcp.XcpAndroidApp.Fragments.CalFragment.OnViewCreated (Android.Views.View,Android.OS.Bundle) [0x0000a] in W:\DevWA\MobileXCP~BARONDST\MobileXCP\MobileXcp\MobileXcp.XcpAndroidApp\Fragments\CalFragment.cs:56
   at Android.Support.V4.App.Fragment.n_OnViewCreated_Landroid_view_View_Landroid_os_Bundle_ (intptr,intptr,intptr,intptr) <IL 0x0001c, 0x0018a>
   at (wrapper dynamic-method) object.ddd7ad4c-d45c-4b14-a0b3-442e265641e5 (intptr,intptr,intptr,intptr) <IL 0x0001d, 0x00041>
   at (wrapper native-to-managed) object.ddd7ad4c-d45c-4b14-a0b3-442e265641e5 (intptr,intptr,intptr,intptr) <IL 0x00020, 0x000bb>
   at <unknown> <0xffffffff>
   at (wrapper managed-to-native) Java.Interop.NativeMethods.java_interop_jnienv_call_boolean_method (intptr,intptr&,intptr,intptr) <IL 0x0000c, 0x000ad>
   at Java.Interop.JniEnvironment/InstanceMethods.CallBooleanMethod (Java.Interop.JniObjectReference,Java.Interop.JniMethodInfo) [0x0006b] in /Users/builder/data/lanes/3340/4e275588/source/Java.Interop/src/Java.Interop/Java.Interop/JniEnvironment.g.cs:11300
   at Android.Runtime.JNIEnv.CallBooleanMethod (intptr,intptr) [0x00000] in /Users/builder/data/lanes/3340/4e275588/source/monodroid/src/Mono.Android/src/Runtime/JNIEnv.g.cs:114
   at Android.Support.V4.App.FragmentManagerInvoker.ExecutePendingTransactions () <IL 0x00038, 0x00187>
   at MvvmCross.Droid.Support.V4.MvxCachingFragmentPagerAdapter.FinishUpdate (Android.Views.ViewGroup) <IL 0x00022, 0x0013f>
   at Android.Support.V4.View.PagerAdapter.n_FinishUpdate_Landroid_view_ViewGroup_ (intptr,intptr,intptr) <IL 0x00013, 0x0012f>
   at (wrapper dynamic-method) object.66d746e7-6be1-43de-ad2b-4aa1fe86d6d6 (intptr,intptr,intptr) <IL 0x00017, 0x00039>
   at (wrapper native-to-managed) object.66d746e7-6be1-43de-ad2b-4aa1fe86d6d6 (intptr,intptr,intptr) <IL 0x0001f, 0x000b2>
 
 Attempting native Android stacktrace:
  at _ZN7SkPaintaSERKS_+84 [0x7f8417ba2524]
  at _ZN7android5PaintaSERKS0_+24 [0x7f8419f39bb8]
  at ???+24 [0x730ab225]
 
 =================================================================
 Got a SIGSEGV while executing native code. This usually indicates
 a fatal error in the mono runtime or one of the native libraries 
 used by your application.
 =================================================================
```

**Comment by Noires** [comment]
By looking closer at it, it seems my GREF count is increasing a lot. Which might explain why calling GC.Collect() in between switching through the fragments reduces the occurence of the error?

**Comment by Noires** [comment]
Related question to that:
I have a pool of listitems, which get added to different ObservableCollections. What happens to the bindings if the ObservableCollection gets cleared, but the pool and therefore the items in the collection still exist?

Because at the moment i cant reproduce the issue when i just recreate the objects per list, transfer all values and put them in the list without any other connection.

**Comment by kjeremy** [comment]
Can you come up with a repro case for this?  How often does it happen?

**Comment by Noires** [comment]
I will try to make a seperate project for it soonish.

Basically what i got:
1. A Service who prepares a list of entries. (Around 2-5k entries)
2. Two Viewmodels with a ObservableCollection, which gets filled by the data of the service.
3. Two Views which call a Refresh and Update the ObservableCollection by clearing it (basically just clearing the references since the data is still in the service) and refilling it with the data of the service.
The Refresh happens when the view gets shown. One view gets always put on the backstack and the other one gets destroyed when moving back to the old view.

So basically i jump forth and back between the two viewmodels and after a few times a SIGSEV appears, **always** on the bindings. And by logging the grefcount, i saw that it increased alot but didnt went down anymore.

Calling GC() reduced the occurence of this case a little, but not by much, still happened after max 10 switches between the views.

But changing the way the entries get added to the observablecollection helped immensly, if it didnt even solve the problem complety (didnt get a crash yet). So instead of adding the reference of the entries out of the Service, i recreate the entrie and copy all its values and add it then into the observablecollection.

So it seems to me at the moment, that clearing the observable collection with the references, doesnt seem to have the same effects on the bindings when the instance of a new item gets cleared?

**Comment by Noires** [comment]
The repository is up:
https://github.com/Noires/MvxSIGSEVTest

Basically just press the NextView Button and then use the Back Button on your phone, those steps repeating. Switching between the tabs might help to let it occur faster since the Update-Method is called there each time, but isnt the issue here.

Uncommenting the stuff in the Update() method and commenting the previous two lines out resolves the issue by using a extra binder item:
`        public void Update()
        {
            TestList.Clear();
            foreach (var item in m_listService.ListItems)
            {
                TestList.Add(item);
                item.Count = 0;
                // Switch out above with following code to avoid error:
                //var binder = new ListItemBinder(item);
                //TestList.Add(binder);
                //binder.Count = 0;
            }
        }`

I apologize for some confusing namespaces or the appname, since it is based on the MvxRecyclerLeakTest and didnt have much time to spend on cleaning it up.

**Comment by kjeremy** [comment]
I haven't tested this but there's nothing obvious that jumps out at me other than you don't actually check that properties have changed before raising the event... however I doubt that's an issue here.  What's your intuition as to why using `ListItemBinder` works around the issue?

Another question: is this only reproducible on the emulator?  I know that it has a lower GREF limit.  Can you reproduce this on a phyiscal device?

**Comment by Noires** [comment]
In my opinion what happens here is:

1. Case "No Binder"
I take references of the objects from a pool (just another list with the same item type) out of a service
and add them to my individual ViewModel Lists if needed, which are binded to the Views. If i clear the lists now the bindings doesnt get cleared when the reference gets removed out of the ViewModel or atleast something happens that the bindings sometimes still get accessed? Since the real listitems are still existent in the pool in the service and only the references got removed.

Atleast here we always have the SIGSEV Error in SetValueImpl.

2. Case "Binder"
Basically the same case like 1. Case, just instead of adding the references, i create a binding object which saves the reference and just acts as a wrapper object. 
That means if the item in the ViewModel gets removed, the binding item really gets removed instead of just a reference of the item. 

To the question:
Iam only testing on  physical devices and it differs from device to device a little (Tested it on 4 total).

**Comment by kjeremy** [comment]
I'm seeing a crash because `OverViewViewModel2` doesn't have the property: `ShowNextScreen`

**Comment by kjeremy** [comment]
Thanks I can reproduce this.

**Comment by kjeremy** [comment]
If I through in a GC after clearing the list I don't hit this and get: 

> Weak Target is null in MvxTextViewTextTargetBinding - skipping set 

from https://github.com/MvvmCross/MvvmCross/blob/a641897bfd12acb17b6675d50b66c92d8ecd9a95/MvvmCross/Core/Binding/Bindings/Target/MvxConvertingTargetBinding.cs#L34.

My guess is that `Target`: https://github.com/MvvmCross/MvvmCross/blob/a641897bfd12acb17b6675d50b66c92d8ecd9a95/MvvmCross/Core/Binding/Bindings/Target/MvxTargetBinding.cs#L21 is still alive even thought he Java side is dead... or something like that.

**Comment by Noires** [comment]
Glad, that you could reproduce it, since it had driven me nuts, till i figured out why it actually appeared.

Yeah seems that somehow the process of freeing is different, when only a reference of the object gets removed out of the binded list and not the object itself is freed. 

Can only imagine that it is related to the target, since it crashes somewhere after that:

```
protected override void SetValueImpl(object target, object toSet)
        {
            ((TextView)target).Text = (string)toSet;
}
```
at  MvvmCross/MvvmCross/Binding/Droid/Target/MvxTextViewTextTargetBinding.cs for me.

So yeah probably the java sides died, but the binding somehow remained?

Not sure at this point, since i lack insight of how the bindings exactly work here.

**Comment by kjeremy** [comment]
Fun fact: if I call `ToString` on `target` in `SetValueImp` I can't reproduce the crash.  I also went back to `Mvx.RegisterType<IMvxBindingContext, MvxBindingContext>();` to make sure this wasn't a threading issue with the bindings.

**Comment by Noires** [comment]
More information on this:
If i use MvxNotifyPropertyChanged in my listitem, so i can bind it via listitem.PropertyChanged in my ListItemBinder. And just raise there the changed event there aswell (since the properties are the same), i manage to get the same exception than before. 
Basically like that:
```
        public CalCategoryBinder(ICalCategory category)
        {
            Category = category;
            category.PropertyChanged += CategoryOnPropertyChanged;
        }

        private void CategoryOnPropertyChanged(object sender, PropertyChangedEventArgs propertyChangedEventArgs)
        {
            RaisePropertyChanged(propertyChangedEventArgs.PropertyName);
        }
```
If i just do it like that:

```
public string Name { 
get { return Category.Name; } 
set { Category.Name = value; RaisePropertyChanged(() => Name); } 
}
```

I have no SIGSEV, though its a hassle to handle all the property changes myself again.

**Comment by kjeremy** [comment]
I'm not sure what that tells us.

**Comment by Noires** [comment]
Dunno if it really helps at this point. Just stumbled on it.

Sadly #1390 doesnt help here atleast, so just gathering more informations.

**Comment by Noires** [comment]
Okay i tested now more in my own application and was logging how often SetValue in MvxConvertingTargetBinding gets called for a certain value.
Basically by each switch it just got called one more time, which obviously shouldnt be the case, since the bindings should get cleared.
And after seven times switching and therefore prolly 7 bindings should be there, it crashses just inbetween updating all of those.

So just need to figure out why this does happen.

**Comment by kjeremy** [comment]
Are you saying that the event is constantly subscribing and never unsubscribing?  If that's the case if you force a GC with something like `System.GC.Collect(GC.MaxGeneration)` does that "fix" the issue?

**Comment by kjeremy** [comment]
After the following:
1. Upgrading to MvvmCross 4.2.1
2. Fixing the sample project so that MainLauncher was true on the main activity and NOT the splash screen.
3. Adding my own `MvxTextViewTextTargetBinding2` that binds to `Text2`

I get an immediate `ObjectDisposedException`:
```csharp
    protected override void SetValueImpl(object target, object toSet)
        {
            try
            {
                var tv = (TextView)target;
                var val = (string)toSet;

                tv.Text = val;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine(ex.Message); // Catches the exception
            }
        }
```

This really seems to point at #1390.  I'll see if I can get that going.  Oddly enough I didn't see this exception the other day (the runtime just kept crashing).

**Comment by kjeremy** [comment]
Ok tried it out but you are right in that it's not the fix for this issue since it's not the target binding that's the issue it's the damn `View`.

**Comment by kjeremy** [comment]
This has something to do with the ViewPager or TabLayout.  If I comment out the "Test2" `FragmentInfo` in EITHER `OverViewFragment` or `OverViewFragment2` then I can't reproduce the crash:

```csharp
var fragments = new List<MvxCachingFragmentStatePagerAdapter.FragmentInfo>
            {
                // TODO: Remove Strings
                new MvxCachingFragmentStatePagerAdapter.FragmentInfo("Test", typeof (RecyclerViewTestFragment),
                    typeof (RecyclerViewTestViewModel)),
                new MvxCachingFragmentStatePagerAdapter.FragmentInfo("Test2", typeof (RecyclerViewTestFragment2),
                    typeof (RecyclerViewTestViewModel2))
            };
```

**Comment by Noires** [comment]
Yeah, it seems to be constantly subscribing but never to unsubscribe. Will try out to force the GC on monday, since iam lacking a laptop at home at the moment.

And will try to update clean to 4.2.1, but last time i had some weird assembly warnings, which i couldnt get rid off, so i reverted it back.

Hum, thats kinda odd. Didnt actually think it would have anything to do with the ViewPager or TabLayout.

**Comment by kjeremy** [comment]
If I insert a GC into `OverViewFragment` I get
> Weak Target is null in MvxTextViewTextTargetBinding2 - skipping set
Which makes sense except that something is keeping the binding alive.  I suspect it's either a bug in the `MvxRecyclerView` and related classes or something with fragments.  There were a bunch of fragment fixes in support 24.0 but Xamarin won't release that lib for awhile.

**Comment by kjeremy** [comment]
Note: A workaround is to check for `target.Handle == IntPtrz.Zero` before we set values.  But that just hides the problem.

**Comment by kjeremy** [comment]
We don't have this problem with `MvxListView`.  It looks like in the typical case of binding things to say an activity or fragment the bindings are cleared immediately because the binding context is associated with the view during inflation.  How this works is a bit of a mystery to me though.  When the activity/fragment `OnDestroyView` is called we clear the bindings and things get `Disposed` of properly.

The above does not happen with views like `MvxListView` and `MvxRecyclerView` where we hold  adapters where each item view creates it's own binding context.  In cases like these we have to wait for the view the be collected which can take a VERY long time.  In this case `Dispose(false)` is often called.  What that means for event handlers and such I don't know... we may be leaking events.  This is me just writing down what I observe because I'll probably forget it all.

The above is background info.  I don't yet know why we don't see a similar issue (C# is alive, Java is dead) with `MvxListView` yet.  Unfortunately there's no `OnViewDestroyed` for `View`s... the closest we can get is `OnAttachedToWindow`/`OnDetachedFromWindow`.  Maybe we should simply `Dispose` of our `Mvx` Views in `OnDestroyView` or maybe we can chain binding contexts together and at least clear the bindings at once.

**Comment by Noires** [comment]
Well normally that `Dispose(false)` is called should be normal and the events should properly get destroyed, since they are linked to the objects in `MvxFullbinding`. Though it could help to call `Dispose` manually and ensure that the disposed bindings cant be called anymore with a `disposed `flag till they get cleared up by the GC.

Though it would just hide the problem aswell, since it shouldnt be needed?

**Comment by kjeremy** [comment]
Ok! I think I sort of have a fix.

In `MvxRecyclerView`:
`private readonly List<WeakReference<MvxBindingContext>> _viewHolderContexts = new List<WeakReference<MvxBindingContext>>();`

Then in `OnCreateViewHolder` we add to the list and in `OnDetachedFromWindow` we call the following new function on the adapter:
```
        public void Cleanup()
        {
            foreach(var bindingContextRef in _viewHolderContexts)
            {
                MvxBindingContext bindingContext;
                if (bindingContextRef.TryGetTarget(out bindingContext))
                    bindingContext.ClearAllBindings();
            }

            _viewHolderContexts.Clear();
        }
```

This seems to fix the crash for me.  What I don't know yet is how this plays with temporary detachment or when we swap adapters out.  Thoughts?

**Comment by kjeremy** [comment]
Actually instead of child contexts maybe we should track view holders.

**Comment by kjeremy** [comment]
I've submitted a PR that fixes this for me.  Please test.

**Comment by Noires** [comment]
It definitly seems to work, atleast i didnt get a crash yet. Thank you a lot for this!

Sadly i doesnt have any thoughts more thoughts to the emporary detachment or adapter swapping, since i didnt do much in that direction yet~

**Comment by nmilcoff** [comment]
@vrishe would you mind opening a new issue referencing this one?

**Comment by vrishe** [comment]
Sure, @milcoff but as soon as I'll be ready with kind of demo project where this case is consistently reproduced.

At current time I wasn't able to reproduce such a behaviour in my sandbox project. Feels like an issue is local to my main project, rather than framework-bound one. Thus, I might have hurried with reviving this discussion thread at all. =\
